### ISSUE TITLE
NAT rebinding mitigation

### ISSUE THREAD

**Issue by agrover** [mozilla/neqo#551]
@martinthomson

-transport section 8,2 says:

> Effective NAT traversal needs additional synchronization mechanisms that are not provided here.

What exactly should we be doing? Seems like keepalive packets are needed if NAT rebinding time is less than idle timeout.

cc @ddragana

**Comment by martinthomson** [comment]
The point made there is that if you have two endpoints who are both behind a NAT, then you often need some sort of coordination.

Generally, I think that it is better if we keep our idle timeout close to what we believe the NAT rebinding time is.  At least until we are more confident in our ability to manage migration properly.  For instance, ideally we would want to rotate connection ID if we were idle for >30s.

That doesn't mean that we need a keep-alive as you describe.  Connection IDs mean that we don't need keep-alives until close to the idle timeout.  And we only need that if we have reason to keep the connection; better to let the connection drop in most cases.  We don't need to keep the path open.  Though in the absence of a connection ID, we might choose to abandon connections faster or change the logic we use for keep-alives, see below.

What we might do is add an API to the transport that says "keep this connection alive".  HTTP would enable that if it had something to send (even if it were blocked) or it was waiting on a response.  At that point we would send keep-alives (a PING probably) within the idle timeout period to keep the connection alive.  The only tweak to this I would add is that if the peer didn't ask us to use a connection ID, or to avoid active migration, then we might reduce the timer we use to 30s.

**Comment by martinthomson** [comment]
I will open a new issue to track the addition of an API for keeping the connection alive.
